shared/src/test/diff/ex/RepMin.mls

```diff
@@ -0,0 +1,288 @@
+//│ fun walkLazy: forall 'a 'get. (Leaf[Num & 'a] | Node[Num & 'a], {get: Int & 'get}) -> ['a, Lazy[Leaf['get | 'a] | Node['get | 'a]]]
+
+
+// * One way to tie the recursive knot
+
+class repmin(t: Tree[Int]) {
+  val mn_res = walkLazy of t, Lazy(() => mn_res.0)
+  val mn = mn_res.0
+  val tree = mn_res.1.get
+}
+//│ class repmin(t: Tree[Int]) {
+//│   val mn: Int
+//│   val mn_res: [Int, Lazy[Leaf[Int] | Node[Int]]]
+//│   val tree: Leaf[Int] | Node[Int]
+//│ }
+
+let rm = repmin(tree)
+//│ let rm: repmin
+//│ rm
+//│    = repmin {}
+//│ // Output
+//│ Traversing leaf 1
+//│ Traversing leaf 5
+//│ Traversing leaf 3
+//│ Traversing leaf 3
+//│ Traversing leaf 2
+
+[rm.mn, String of rm.mn_res]
+//│ [Int, Str]
+//│ res
+//│     = [
+//│         1,
+//│         '1,Lazy(Node(Node(Node(Leaf(1), Leaf(1)), Leaf(1)), Node(Leaf(1), Leaf(2))))'
+//│       ]
+
+String of rm.tree
+//│ Str
+//│ res
+//│     = 'Node(Node(Node(Leaf(1), Leaf(1)), Leaf(1)), Node(Leaf(1), Leaf(2)))'
+
+
+// * Another way
+
+fun repmin(t) =
+  module Rec {
+    val mn_res = walkLazy of t, Lazy(() => mn_res.0)
+  }
+  [Rec.mn_res.0, Rec.mn_res.1.get]
+//│ fun repmin: forall 'a. (Leaf[Num & 'a] | Node[Num & 'a]) -> ['a, Leaf['a] | Node['a]]
+
+// :Fuel 1000000
+// FIXME why does it run out of fuel?!
+String of repmin(tree)
+//│ ╔══[ERROR] Subtyping constraint of the form `forall ?a ?b ?get. ?b -> [?a, ?get] <: ?c -> ?d` took too many steps and ran out of fuel (10000)
+//│ ║  l.197: 	String of repmin(tree)
+//│ ║         	          ^^^^^^^^^^^^
+//│ ╙── Note: use flag `:ex` to see internal error info.
+//│ Str
+//│ res
+//│     = '1,Node(Node(Node(Leaf(1), Leaf(1)), Leaf(1)), Node(Leaf(1), Leaf(2)))'
+//│ // Output
+//│ Traversing leaf 1
+//│ Traversing leaf 5
+//│ Traversing leaf 3
+//│ Traversing leaf 3
+//│ Traversing leaf 2
+
+
+// * A way that could work but currently doesn't because of a codegen limitation
+
+:ge
+fun repmin(t) =
+  let rec mn_res = walkLazy of t, Lazy(() => mn_res.0)
+  [mn_res.0, mn_res.1.get]
+//│ fun repmin: forall 'a. (Leaf[Int & 'a] | Node[Int & 'a]) -> ['a, Leaf['a] | Node['a]]
+//│ Code generation encountered an error:
+//│   unguarded recursive use of by-value binding mn_res
+
+:re
+String of repmin(tree)
+//│ Str
+//│ res
+//│ Runtime error:
+//│   ReferenceError: repmin2 is not defined
+
+
+// * An alternative solution using mutation and thunks instead of laziness
+
+let IntMax = 2147483648
+//│ let IntMax: 2147483648
+//│ IntMax
+//│        = 2147483648
+
+fun repmin(t) =
+  mut let min = IntMax
+  let rec go(t) = if t is
+    Leaf(n) then
+      log of join of "Traversing leaf ", String of n
+      if n < min then set min = n else ()
+      () => Leaf of if n > 2 * min then min else n
+    Node(l, r) then
+      let l' = go(l)
+      let r' = go(r)
+      () => Node of l'(), r'()
+  let res = go(t)
+  [min, res()]
+//│ fun repmin: forall 'min. (Leaf[Int & 'min] | Node[Int & 'min]) -> [2147483648, Leaf[2147483648 | 'min] | Node[2147483648 | 'min]]
```

Comment:
I just noticed the `[2147483648,` part of the inferred type is wrong.

Reply:
Minimized:
```fs
fun repmin(t) =
  mut let min = 123
  let go(t) = set min = t.value
  go(t)
  min
//│ fun repmin: {value: anything} -> 123
```

Reply:
Lol the problem is that let bindings were still generalized even when mutable. Trying to solve this in a principled way now.